fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/impl/datagen/SoundTypeBuilderImpl.java

```diff
@@ -0,0 +1,217 @@
+			}
+		}
+
+		return new SoundType(sounds, category, Optional.ofNullable(subtitle));
+	}
+
+	public record SoundType(List<Entry> sounds, SoundCategory category, Optional<String> subtitle) {
+		private static final Map<String, SoundCategory> CATEGORIES = Arrays.stream(SoundCategory.values()).collect(Collectors.toMap(SoundCategory::getName, Function.identity()));
+		private static final Codec<SoundCategory> SOUND_CATEGORY_CODEC = Codec.stringResolver(SoundCategory::getName, name -> CATEGORIES.getOrDefault(name.toLowerCase(Locale.ROOT), SoundCategory.NEUTRAL));
+		public static final Codec<SoundType> CODEC = RecordCodecBuilder.create(i -> i.group(
+				Entry.CODEC.listOf().fieldOf("sounds").forGetter(SoundType::sounds),
+				SOUND_CATEGORY_CODEC.fieldOf("category").forGetter(SoundType::category),
+				Codec.STRING.optionalFieldOf("subtitle").forGetter(SoundType::subtitle)
+		).apply(i, SoundType::new));
+	}
+
+	private record Entry(Identifier name, RegistrationType type, float volume, float pitch, int weight, int attenuationDistance, boolean stream, boolean preload) {
+		private static final Codec<Entry> MAP_CODEC = RecordCodecBuilder.create(i -> i.group(
+				Identifier.CODEC.fieldOf("name").forGetter(Entry::name),
+				RegistrationType.CODEC.optionalFieldOf("type", RegistrationType.FILE).forGetter(Entry::type),
+				Codec.FLOAT.optionalFieldOf("volume", 1F).forGetter(Entry::volume),
+				Codec.FLOAT.optionalFieldOf("pitch", 1F).forGetter(Entry::pitch),
+				Codec.INT.optionalFieldOf("weight", 1).forGetter(Entry::weight),
+				Codec.INT.optionalFieldOf("attenuation_distance", 16).forGetter(Entry::attenuationDistance),
+				Codec.BOOL.optionalFieldOf("stream", false).forGetter(Entry::stream),
+				Codec.BOOL.optionalFieldOf("preload", false).forGetter(Entry::preload)
+		).apply(i, Entry::new));
+
+		private static final Codec<Entry> STRING_CODEC = Identifier.CODEC.xmap(
+				id -> new Entry(id, RegistrationType.FILE, 1F, 1F, 1, 16, false, false),
+				Entry::name
+		);
+		private static final Codec<Entry> CODEC = Codec.xor(STRING_CODEC, MAP_CODEC).xmap(Either::unwrap, sound -> {
+			if (sound.type() != RegistrationType.FILE
+					|| sound.volume() != 1F
+					|| sound.pitch() != 1F
+					|| sound.weight() != 1
+					|| sound.attenuationDistance() != 16
+					|| sound.stream()
+					|| sound.preload()) {
+				return Either.right(sound);
+			}
+
+			return Either.left(sound);
+		});
+	}
+
+	public static final class EntryBuilderImpl implements EntryBuilder {
+		private final Identifier name;
+		private final RegistrationType type;
+
+		private float volume = 1F;
+		private float pitch = 1F;
+		private int attenuationDistance = 16;
+		private int weight = 1;
+		private boolean stream = false;
+		private boolean preload = false;
+
+		private EntryBuilderImpl(RegistrationType type, Identifier name) {
+			this.type = type;
+			this.name = name;
+		}
+
+		public static EntryBuilder builder(RegistrationType type, Identifier name) {
```

Comment:
Should just expose ctor

Reply:
Generally other builders do not expose the constructor.